---
output: html_notebook
---

####*Objective 9. Perform Gene Set Enrichment Analysis (GSEA)*

<br/>
Packages used in this notebook: `gage`, `pathview`. First verify that they are installed and loaded.
```{r message=FALSE, echo=FALSE}
## Defining install_load function - adapted from: https://gitlab.com/iembry/install.load
install_load <- function(package1, ...) {   
   # convert arguments to vector
   packages <- c(package1, ...)
   # start loop to determine if each package is installed
   for(package in packages) {
       # if package is installed, just load it
       if(package %in% (BiocManager::available()))
          do.call('library', list(package))
       # if package is not installed locally, download, then load
       else {
          BiocManager::install(package)
          do.call("library", list(package))
       }
   }    
}
install_load("gage","pathview")
```


<br/>
Set the directory where you want the results to be stored (this assumes that you have a folder called _RNAseq_ on your desktop, with a subfolder called _Results_).

```{r setup}
## if you ARE using an R notebook (if not the comment out this line):
knitr::opts_knit$set(root.dir = "~/Desktop/RNAseq/Results/") 

## if you ARE NOT using an R notebook (un-comment this line):
#setwd("~/Desktop/RNAseq/Results/") 
```


<br/>
The `limma` package can test for enrichment with [Gene Ontology](http://geneontology.org/) (GO) terms with the `goana()` function. It can also test for [Kyoto Encyclopedia of Genes and Genomes](https://www.genome.jp/kegg/) (KEGG) pathways with the `kegga()` function. But some of the options don't work, and there are not as many as the functions from the `gage` package. So we will use `gage` and then view pathways with `pathview`.    

See the vignettes for `gage` (run `vignette(package="gage")`) and also this R-bloggers [tutorial](https://www.r-bloggers.com/tutorial-rna-seq-differential-expression-pathway-analysis-with-sailfish-deseq2-gage-and-pathview/).


<br/>

The first input to the `gage()` function this is a named vector of log2(fold-change) values (or contrast coefficients), where the names are Entrez IDs. Let's start by doing a gene set analysis for LP vs. ML. To get these you can use the LPvsML coefficients from the efit coefficient matrix. The row names are the Entrez IDs.  

Note: you should look through the help file for `gage()` in addition to the vignette for more information about the methods/options.  

```{r}
?gage
head(EFITE$coefficients)

fold_lpml <- efit$coefficients[, "LPvsML"]
head(fold_lpml) ## Entrez IDs are retained from the coefficient matrix
```

</br>

##### Test for KEGG pathway enrichment  

Lets's test for KEGG pathway enrichment in the LP vs. ML samples. 

After getting the fold changes (coefficients) as above, the next thing we need to do is to get the latest mouse genome KEGG pathways to test.
```{r}
?kegg.gsets

kegg_mouse <- kegg.gsets(species="mmu", id.type="kegg")
names(kegg_mouse)

head(kegg_mouse$kg.sets, n=3) ## look at first 3
```
<br/>
Now we can test whether any of the mouse KEGG pathways are significantly perterbed in the LP vs. ML samples. Using the same.dir option gives us separate lists of pathways that are enriched with up-regulated and down-regulated genes in LP vs. ML.

```{r}
?gage

kegg_all = gage(exprs=fold_lpml, gsets=kegg_mouse$kg.sets, same.dir=TRUE)
```

Let's look at the stats for pathways that are enriched with up-regulated genes in LP vs. ML, ("greater") and those that are enriched with down-regulated genes ("less").
```{r}
lapply(kegg_all, head)
```

The statistics output in the tables are also described in the help file for the `gage()` function. The stat.mean indicates the direction and magnitude of the gene set perterbation, q.val is the adjusted significance level (false discovery rate - you can use 0.1 for example), and set.size represents the number of genes in the pathway.  


</br>

##### Test for KEGG metabolism pathway enrichment  

You can also test the subset of KEGG signaling, metabolism, and disease pathways separately. The full set that we created above with the `kegg.gsets()` function (we named this kegg_mouse) can be subset for either one using indices. For example let's say we want to test if any metabolic pathways are perturbed - we'll start by subsetting these. 

```{r}
names(kegg_mouse)
head(kegg_mouse$met.idx) ## these are indices of the metabolic pathways
kegg_mouse$kg.sets[1] ## this is the first metabolic pathway (index=1)
```

```{r}
kegg_met <- kegg_mouse$kg.sets[kegg_mouse$met.idx] ## all metabolic pathways
```

Make sure the number of pathways matches the number of metabolic indices.
```{r}
length(kegg_mouse$met.idx)
length(kegg_met)
```

Now we can use the kegg_met subset on the LP vs. ML fold changes to find metabolic pathways enriched with up- and down-regulated genes. Say we are interested only in the genes that are down-regulated in LP vs. ML.

```{r}
kegg_met = gage(exprs=fold_lpml, gsets=kegg_met, same.dir=TRUE)
names(kegg_met) 
head(kegg_met$less, n=3)
```


</br>

##### Select and visualize specific pathways 

Now let's pull out the one metabolic pathway that is enriched with genes that are down-regulated in LP vs. ML at FDR<0.1 (Oxidative phosphorylation). We are going to use the `pathview` package to look at it. The input to this requires a KEGG ID, so we will extract this from the results.     

```{r}
## extract KEGG ID and pathway name (first row name of the kegg_met$less data)
op <- rownames(kegg_met$less)[1]
op

## extract the KEGG ID (this code removes everything after the first space from op)
op_kegg <- sub(" .*$", "", op)
op_kegg
```

<br/>

Finally, we can plot this pathway. We feed it the fold changes, pathway KEGG ID, and species. The `pathway()` function automatically creates .png and .xml files, and the genes are highlighted using the fold-change values.    

```{r message=FALSE}
?pathview

pathview(gene.data=fold_lpml, pathway.id=op_kegg, species="mmu")
```
