---
output: html_notebook
---

####*Objective 1. Download RNA-seq files from GEO*


<br/>
Packages used in this notebook: `GEOquery`, `edgeR`. First verify that these are installed and loaded.
```{r message=FALSE, echo=FALSE}
## Defining install_load function - adapted from: https://gitlab.com/iembry/install.load
install_load <- function(package1, ...) {   
   # convert arguments to vector
   packages <- c(package1, ...)
   # start loop to determine if each package is installed
   for(package in packages) {
       # if package is installed, just load it
       if(package %in% (BiocManager::available()))
          do.call('library', list(package))
       # if package is not installed locally, download, then load
       else {
          BiocManager::install(package)
          do.call("library", list(package))
       }
   }    
}
install_load("GEOquery", "edgeR")
```

<br/>
Set the directory where you want the data to be stored (this assumes that you have a folder called _RNAseq_ on your desktop, with a subfolder called _Data_). 
```{r setup}
## if you ARE using an R notebook (if not the comment out this line):
knitr::opts_knit$set(root.dir = "~/Desktop/RNAseq/Data/") 

## if you ARE NOT using an R notebook (un-comment this line):
#setwd("~/Desktop/RNAseq/Data/") 
```


<br/>
The files we will be using for this are from the GEO database, accession # [GSE63310](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63310):


<br/>
Retreive the raw RNA-seq files using the GEO accession. Usually these are either provided in CEL format (Affymetrix arrays), or other formats such as plain TXT files (other platforms). In this case we will be using TXT files.  
```{r include=FALSE}
?getGEOSuppFiles

gse <- getGEOSuppFiles("GSE63310")
```

<br/>
Note that this created a sub-directory with the accession name and downloaded a .tar file, which we will unpack into our main working Data directory.
```{r}
untar("./GSE63310/GSE63310_RAW.tar")
```

<br/>
We will be analyzing 9 of the 11 samples, all in separate files. 
```{r}
files <- c("GSM1545535_10_6_5_11.txt", "GSM1545536_9_6_5_11.txt", "GSM1545538_purep53.txt", "GSM1545539_JMS8-2.txt", "GSM1545540_JMS8-3.txt", "GSM1545541_JMS8-4.txt", "GSM1545542_JMS8-5.txt", "GSM1545544_JMS9-P7c.txt", "GSM1545545_JMS9-P8c.txt")
```
Let's de-compress these.  
```{r}
for(i in paste(files, ".gz", sep="")) {
    gunzip(i, overwrite=TRUE)
}
```

<br/>
We could open any of these in a text editor, but let's read the first file into a data frame so we can take a look at it in R.
```{r}
files[1]
f1 <- read.delim(files[1])
dim(f1)
head(f1)
```
As we can see, this dataframe has three columns: the gene ID (Entrez), length of the gene, and read counts from the run.

<br/>
Now we'll create a DGEList object which will contain data from all files. This requires us to indicate which columns represent the gene names and raw read counts. This assumes there is only one gene per row (no duplicates).
Note: DGE = digital gene expression. 
```{r}
?readDGE

dge <- readDGE(files, columns=c(1,3))
dge
```








