---
output: html_notebook
---

####*Objective 2. Annotate the samples*

<br/>
Packages used in this notebook: `edgeR`, `org.Mm.eg.db`. First verify that these are installed and loaded.
```{r message=FALSE, echo=FALSE}
## Defining install_load function - adapted from: https://gitlab.com/iembry/install.load
install_load <- function(package1, ...) {   
   # convert arguments to vector
   packages <- c(package1, ...)
   # start loop to determine if each package is installed
   for(package in packages) {
       # if package is installed, just load it
       if(package %in% (BiocManager::available()))
          do.call('library', list(package))
       # if package is not installed locally, download, then load
       else {
          BiocManager::install(package)
          do.call("library", list(package))
       }
   }    
}
install_load("edgeR","org.Mm.eg.db")
```

<br/>
Set the directory where you have the data to be stored (this assumes that you have it in a folder called _RNAseq_ on your desktop, with a subfolder called _Data_). 
```{r setup}
## if you ARE using an R notebook (if not the comment out this line):
knitr::opts_knit$set(root.dir = "~/Desktop/RNAseq/Data/") 

## if you ARE NOT using an R notebook (un-comment this line):
#setwd("~/Desktop/RNAseq/Data/") 
```


<br/>
First, let's read in all of the files that we left out the first time around. In Objective 1 we created a vector of filenames stored in the `files` object. It is assumed that this object is still in your environment. If it is not found then you need to re-create it. 
```{r}
files
```

<br/>
The `readDGE()` function from the `edgeR` package conveniently reads all of the files at once. Recall that we only want the first and third columns (gene names and counts, respectively).
```{r}
dge <- readDGE(files, columns=c(1,3))
dge
```

<br/>
Now we need to indicate what the different sample groups are. In this case we'll be able to identify these groups by looking at the Samples section in GEO ([GSE63310](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63310)). In our own experimental data, we'd already have this information by virtue of designing the experiment ourselves.

In this case we have 3 groups: LP (luminal progenitor), ML (mature luminal), and Basal epithelial cells.

<br/>
Now we'll add the sample annotations to the group column of the DGEList object `dge` that we created in the previous section.
```{r}
dge$samples$group <- as.factor(c("LP", "ML", "Basal", "Basal", "ML", "LP", "Basal", "ML", "LP"))
dge$samples
```

<br/>
The counts part of the DGEList contains gene names (ENTREZIDs) in rows and sample names in columns.
```{r}
head(dge$counts)
dge_ids <- rownames(dge)
print("")
print("Entrez IDs:")
head(dge_ids)
print("Number of elements (gene IDs):")
length(dge_ids)
```

<br/>
Now say we want to get gene SYMBOLS as well - this means we'd need the mouse genome. We're able to do this by using the mouse genome info this is built into the `AnnotationDbi` package that comes with Bioconductor. Note the use of `keys=dge_ids` here (we wouldn't want to type them all out since this is a long vector).
```{r}
gene_ids <- select(org.Mm.eg.db, 
    keys=dge_ids, 
    keytype="ENTREZID",
    columns=c("SYMBOL"))
head(gene_ids)
```

<br/>
Now we can add these identifiers to the DGEList object in addition to the samples and counts.
```{r}
names(dge)
dge$gene_ids <- gene_ids
```

```{r}
names(dge)
```
