---
output: html_notebook
---

####*Objective 6. Perform unsupervised clustering of expression data*

<br/>
Packages used in this notebook: `limma`, `RColorBrewer`. First verify that they are installed and loaded.
```{r message=FALSE, echo=FALSE}
## Defining install_load function - adapted from: https://gitlab.com/iembry/install.load
install_load <- function(package1, ...) {   
   # convert arguments to vector
   packages <- c(package1, ...)
   # start loop to determine if each package is installed
   for(package in packages) {
       # if package is installed, just load it
       if(package %in% (BiocManager::available()))
          do.call('library', list(package))
       # if package is not installed locally, download, then load
       else {
          BiocManager::install(package)
          do.call("library", list(package))
       }
   }    
}
install_load("limma","RColorBrewer")
```

<br/>
Set the directory where you want the results to be stored (this assumes that you have a folder called _RNAseq_ on your desktop, with a subfolder called _Results_). 
```{r setup}
## if you ARE using an R notebook (if not the comment out this line):
knitr::opts_knit$set(root.dir = "~/Desktop/RNAseq/Results/") 

## if you ARE NOT using an R notebook (un-comment this line):
#setwd("~/Desktop/RNAseq/Results/") 
```

<br/>
Now we will visualize differences between groups of samples. If it appears that there is some clustering of samples, then doing differential expression analysis can follow to formally test these differences.

<br/>
We'll start by assigning different colors to each group of samples (LP, ML, and Basal) in a similar way to what we did in Objective 4.
```{r}
grps <- dge_sub$samples$group
col.grp <- grps
levels(col.grp) <-  brewer.pal(nlevels(col.grp), "Set1")
col.grp <- as.character(col.grp)
col.grp 
grps
```

<br/>
The `plotMDS()` function from the `limma` package shows us a scatterplot where distances between samples represent typical log2(Fold-Change) between the samples for genes that distinguis the samples (see Details in Help file and in Ritchie ME, Phipson B, Wu D, et al., Nucelic Acids Research 2015). MDS=Multidimensional scaling. You can see variance within and between samples or batches.  

Let's save this plot in a .png file too (can use the `png()` function followed by the plot commands and then `dev.off()`, or you can save it directly from the Plots tab in RStudio).
```{r}
?plotMDS

head(log2cpm_norm)

#png("MDSplot.png")
par(mfrow=c(1,1))
plotMDS(log2cpm_norm, 
    labels=grps, 
    col=col.grp,
    xlab="Log2(Fold-Change)",
    ylab="Log2(Fold-Change)")
title(main="Sample groups")
#dev.off()
```
**Discussion:** What is this plot telling us?

