---
output: html_notebook
---

####*Objective 8. Create heatmap and volcano plots*

<br/>
Packages used in this notebook: `limma`, `gplots`, `RColorBrewer`. First verify that they are installed and loaded.
```{r message=FALSE, echo=FALSE}
## Defining install_load function - adapted from: https://gitlab.com/iembry/install.load
install_load <- function(package1, ...) {   
   # convert arguments to vector
   packages <- c(package1, ...)
   # start loop to determine if each package is installed
   for(package in packages) {
       # if package is installed, just load it
       if(package %in% (BiocManager::available()))
          do.call('library', list(package))
       # if package is not installed locally, download, then load
       else {
          BiocManager::install(package)
          do.call("library", list(package))
       }
   }    
}
install_load("limma","gplots","RColorBrewer")
```

<br/>
Set the directory where you want the results to be stored (this assumes that you have a folder called _RNAseq_ on your desktop, with a subfolder called _Results_). 
```{r setup}
## if you ARE using an R notebook (if not the comment out this line):
knitr::opts_knit$set(root.dir = "~/Desktop/RNAseq/Results/") 

## if you ARE NOT using an R notebook (un-comment this line):
#setwd("~/Desktop/RNAseq/Results/") 
```

Now that we've done some analysis of the data, we'd ought to consider summarizing our findings in a visual format. While figures may not convey the depth of information that quantitive outputs might display, it does make our results far easier to interpret at a glance. Here we'll do a quick survey of some of the more popular RNA-seq visualizations.

</br>

##### Hierarchical clustering with heatmaps:
source: https://combine-australia.github.io/RNAseq-R/06-rnaseq-day1.html

Hierarchical clustering can be a potent way to examine differences between samples, and heatmaps can succintly visualise the clustering of your samples. We can do this using the `heatmap.2()` function from the `gplots` package. In this example `heatmap.2()` calculates a matrix of euclidean distances from the logCPM (logcounts object) for the 500 most variable genes. (Note this has more complicated code than plotting principle components using plotMDS).

You'll notice that we're using the `RColorBrewer` package. This allows us to refine the aesthetics of our figures by using the `brewer.pal()` function to select different color schemes. “RdYlBu” is a common choice, and “Spectral” is also nice.

Let’s select data for the 500 most variable genes and plot the heatmap. First we'll estimate the variance for each row in the log2cpm_norm matrix.
```{r}
?apply
var_genes <- apply(log2cpm_norm, 1, var)
head(var_genes)
```
Then we get the gene names for the top 500 most variable genes.
```{r}
select_var <- names(sort(var_genes, decreasing=TRUE))[1:500]
head(select_var)
```
And then we'll subset the log2cpm_norm matrix to those top 500 variable genes.
```{r}
highly_variable_lcpm <- log2cpm_norm[select_var,]
dim(highly_variable_lcpm)
head(highly_variable_lcpm)
```
Now we'll plot the heatmap.
```{r}
## Get some nicer colors
## to display brewer colors (use diverging colors in this instance): display.brewer.all()
mypalette <- brewer.pal(11,"RdBu") 
morecols <- colorRampPalette(mypalette)
# Set up color vector for celltype variable
col.cell <- c("honeydew4", "lightblue3", "lightcyan2")[c(dge$samples$group)]

# Plot the heatmap
png(filename="heatmap.png")
heatmap <- heatmap.2(highly_variable_lcpm, col=rev(morecols(50)), trace="none",
                    main="Top 500 most variable genes - ML v. LP v. Basal",
                    ColSideColors=col.cell,scale="row",
                    srtCol = 25,
                    labRow = NA)
dev.off()

plot(heatmap$colDendrogram)
plot(heatmap$rowDendrogram)
```
This gives us a good overview of just how much variance is occuring between the LP, MP, and Basal groups. We can even see that in the "colDendogram" our LP and MP samples cluster together while the Basal group is split from those groups at a relatively high level. This makes the distance plots we saw earlier make a little more sense as we can clearly see that the LP and MP groups are much more similar to one another than the Basal group.

That said, it is admittedly a little difficult to select specific genes for analysis just from this... so let's pull out groups of interest from the heatmap:
```{r}
top_500_clustered <- as.hclust(heatmap$rowDendrogram)
```

Creating dataframe of top 500 most variable genes, ranked and sorted:
```{r}
# define the clusters (essentially picking where to "cut"" the dendogram and select clusters)
# cutree() returns a vector of cluster membership in the order of the original data rows
mycl <- cutree(top_500_clustered, h=max(40))

# examine it
#mycl

# examine the head and tail of cluster membership by it's order in the heatmap
head(mycl[top_500_clustered$order], n=10)
tail(mycl[top_500_clustered$order], n=10)

# you could either grab a cluster
#cluster1 <- highly_variable_lcpm[mycl == 1,]

# or simply add the cluster ID to your data. This is what we'll do.
cldat <- cbind(highly_variable_lcpm, clusterID=mycl)

# examine the data with cluster ids attached, and ordered like the heat map
top_diff_genes <- as.data.frame(cldat[top_500_clustered$order,])

# check is gene IDs table is still in memory:
head(gene_ids)

#use gene IDs to rename genes in hr...
top_diff_genes <- tibble::rownames_to_column(top_diff_genes, "ENTREZID")
top_diff_genes <- merge(gene_ids, top_diff_genes, by="ENTREZID")
head(top_diff_genes, n=10)
```

There are a few null ENTREZ IDs we'd need to look up, but at least now we're getting a better idea of exactly which genes specifically account for the most variance between our samples. That said, there are many parameters to play with here...

For instance, we may only be interested in observing genes contained in cluster #1:
```{r}
mycl <- cutree(top_500_clustered, h=max(40))
head(mycl[top_500_clustered$order], n=10)
tail(mycl[top_500_clustered$order], n=10)

# Now we'll grab a cluster
cluster1 <- highly_variable_lcpm[mycl == 1,]
# as opposed to adding the cluster ID to the data, as we did previously.
#cldat <- cbind(highly_variable_lcpm, clusterID=mycl)

top_diff_genes <- as.data.frame(cluster1)

#use gene IDs to rename genes in hr...
top_diff_genes <- tibble::rownames_to_column(top_diff_genes, "ENTREZID")
top_diff_genes <- merge(gene_ids, top_diff_genes, by="ENTREZID")
head(top_diff_genes, n=10)
```

And if we are interested in further dissecting genes by cluster, we may adjust where that threshold is as follows:
```{r}
mycl <- cutree(top_500_clustered,
               h=max(20) # here is the line to edit in this case - sets the height at which to "cut" the dendogram
               )

head(mycl[top_500_clustered$order], n=10)
tail(mycl[top_500_clustered$order], n=10)

# This time we're adding the cluster assignment to the data to observe what this threshold adjustment actually accomplished
cldat <- cbind(highly_variable_lcpm, clusterID=mycl)

top_diff_genes <- as.data.frame(cldat[top_500_clustered$order,])
top_diff_genes <- tibble::rownames_to_column(top_diff_genes, "ENTREZID")
top_diff_genes <- merge(gene_ids, top_diff_genes, by="ENTREZID", sort = F)
head(top_diff_genes)
```
Here we're "eyeballing" our cutoff and aimed to split our data into 5 different clusters based on the heatmap. There are empirical methods to determine the ideal number of clusters based on the properties of your data - more info can be found here:
https://www.datanovia.com/en/lessons/determining-the-optimal-number-of-clusters-3-must-know-methods/#average-silhouette-method

To conclude this heatmap section, the following page is an excellent resource that details some of the more specific functionality of `heatmap.2()`:
https://bioinfo.umassmed.edu/bootstrappers/bootstrappers-courses/pastCourses/rCourse_2014-09/Session_6/Session_6.html


</br>

#####Volcano Plots

Another way to visualize differential gene experession is via volcano plots. Here we'll quickly rehash how to get your data formatted appropriately for use with the `limma` package's function `volcanoplot()` and plot a simple example.
```{r}
### Rehashing how to format data for volano plots (from notebooks 6, 7, 8)
# start with our groups
grps # this variable contains our groups of samples (ML, LP, Basal) - created in notebook #7 (unsupervised clustering)
# creating design matrix (dummy variable creation for each sample and the group it belongs to)
design <- model.matrix(~0 + grps)
# relabeling columns
colnames(design) <- gsub("grps", "", colnames(design))
#creating contrast matrix
contrast <- makeContrasts(
   BasalvsLP = Basal - LP, 
   BasalvsML = Basal - ML, 
   LPvsML = LP - ML, 
   levels = colnames(design))
# Now we use `voom()` from `limma` to combine log2(CPM) values from the filtered, normalized data with the design matrix
dge_voom <- voom(dge_sub_norm, design)
# Using limma's lmfit() function - this is the linear model fit step
dge_fit <- lmFit(dge_voom, design)
# Using contrast matrix to compare the groups
grp_fit <- contrasts.fit(dge_fit, contrasts=contrast)
# Empirical Bayes method (limma) to use information across all genes - this should make the residual variances independent of the expression levels.
efit <- eBayes(grp_fit)
```

Now we'll create a simple plot using the `volcanoplot()` function - for each comparison we'll label the top 10 differentially expressed genes.

First we'll replace the Entrez IDs with gene symbols for easier interpretation:
```{r}
# convert gene IDs to Gene Symbols for easier interpretation
efit_genes <- names(efit$Amean)
efit_genes <- as.data.frame.AsIs(efit_genes)
colnames(efit_genes) <- "ENTREZID"
efit_genes <- merge(gene_ids, efit_genes, by="ENTREZID", sort = F)
names(efit$Amean) <- efit_genes$SYMBOL
```

And now we're ready to plot:
```{r}
?volcanoplot
# basal vs lp
volcanoplot(efit,
            coef = 1,
            style = "p-value",
            highlight = 10,
            names = names(efit$Amean),
            hl.col="blue",
            main = "Basal vs LP",
            xlab = "Log2 Fold Change",
            ylab = NULL,
            pch=20,
            cex=0.35)

# basal vs ml
volcanoplot(efit, coef = 2, style = "B-statistic", highlight = 10, names = names(efit$Amean), 
            hl.col="blue",
            main = "Basal vs ML", xlab = "Log2 Fold Change", ylab = NULL, pch=20, cex=0.35)

# lp vs ml
volcanoplot(efit,coef = 3, style = "p-value", highlight = 10, names = names(efit$Amean), 
            hl.col="blue", 
            main = "LP vs ML", xlab = "Log2 Fold Change", ylab = NULL, pch=20, cex=0.35)
```
This is a succinct way to visualize differential expression patterns, but let's pull out the labeled genes for further observation.

```{r}
?topTable
# top hits:
BasalvLP_top <- topTable(efit, coef = 1, number = 10, sort.by = 'P')
BasalvLP_top

# top hits:
BasalvML_top <- topTable(efit, coef = 2, number = 10, sort.by = 'P')
BasalvML_top

# top hits:
ML_v_LP_top <- topTable(efit, coef = 3, number = 10, sort.by = 'P')
ML_v_LP_top
```
Here we have pulled out the top 10 most significant differentially expressed genes that were labeled on each volcano plot. With certain datasets the row names will be gene symbols. In this case, however, we will need to process these row names to replace them with the gene symbols.

```{r}
# convert gene IDs to Gene Symbols for easier interpretation
ML_v_LP_top <- tibble::rownames_to_column(ML_v_LP_top, "ENTREZID")
ML_v_LP_top <- merge(gene_ids, ML_v_LP_top, by="ENTREZID", sort = F)
ML_v_LP_top
```


</br>

##### Some notes on volcano plots:  
It is possible to refine this figure for publication within `limma `, but we'd recommend using the EnhancedVolcano package to generate publication-ready plots.

In the interest of time, we won't be covering the EnhancedVolcano package for highly customizable plots. We will be introducing more advanced RNA-Seq content soon and this planned to be part of those notebooks. If you wish to give this a try yourself, see the following bioconductor vignette.

https://bioconductor.org/packages/devel/bioc/vignettes/EnhancedVolcano/inst/doc/EnhancedVolcano.html

Fair warning: it will take some data cleaning and restructuring to implement. If you'd like us at the DSI to take care of that for you, please let us know after the workshop. We will happily send you the advanced materials once they have been prepared and validated.